File: leetcode/p1022/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private int sum;
    public int sumRootToLeaf(TreeNode root) {
        sum = 0;
        dfs(root, 0);
        return sum;
    }
    
    private void dfs(TreeNode node, int num) {
        if (node == null) {
            return;
        }
        
        num = (num << 1) + node.val;
        if (node.left == null && node.right == null) {
            sum += num;
            return;
        }
        
        dfs(node.left, num);
        dfs(node.right, num);
    }
}

File: leetcode/p1022/Solution.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    private var sum = 0
    fun sumRootToLeaf(root: TreeNode?): Int {
        sum = 0
        dfs(root, 0)
        return sum
    }
    
    private fun dfs(node: TreeNode?, num: Int) {
        if (node == null) {
            return
        }
        
        val curNum = (num shl 1) + node.`val`
        if (node.left == null && node.right == null) {
            sum += curNum
            return;
        }
        dfs(node.left, curNum)
        dfs(node.right, curNum)
    }
}
